### PR TITLE
FEATURE: SD-889 typing for export, add filter, add widget on kd

### PR DESCRIPTION
**Pull request by lvkhanh**
- Release Type: **patch**
- Related PR: https://github.com/gooddata/gdc-dashboards/pull/2819

### CODE REVIEW

**Comment by yenkins** [comment]
Currently destined for [release R206](https://confluence.intgdc.com/display/Dev/R206).
This release is **zero downtime** release.
Code drop: 2020-06-04 00:00.

**Comment by yenkins** [comment]
Story [SD-889](http://jira.intgdc.com/browse/SD-889) KD - Implement to post events in response to GDC application updated
	Status: Hacking
	Priority: Major
	Assignee: Khanh Le

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/98/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/183/ : SUCCESS in 52s

File: src/embedding/kd.ts [hunk 1]
```diff
@@ -308,6 +340,110 @@ export namespace EmbeddedKpiDashboard {
...
+     * -  if KD is in view mode or not showing any dashboard, then CommandFailed will be posted
+     *
+     * @remarks use {@link AddWidgetCommand} to create instances of AddWidget
+     */
+    export type AddWidgetCommand = IGdcKdMessageEvent<GdcKdCommandType.AddWidget, IAddWidgetBody>;
+
+    export type AddWidgetCommandData = IGdcKdMessageEnvelope<GdcKdCommandType.AddWidget, IAddWidgetBody>;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link isAddWidgetCommandData}.
```
Comment (lanhq) [inline src/embedding/kd.ts:390]: `{@link isAddWidgetCommandData}` => `{@link AddWidgetCommandData}`

File: src/embedding/kd.ts [hunk 2]
```diff
@@ -308,6 +340,110 @@ export namespace EmbeddedKpiDashboard {
...
+     * -  if KD shwows dashboard in edit mode or not not showing any dashboard, CommandFailed will
+     *    be posted
+     * @remarks use {@link ExportToPDFCommand} factory function to instantiate
+     */
+    export type ExportToPDFCommand = IGdcKdMessageEvent<GdcKdCommandType.ExportToPDF, null>;
+
+    export type ExportToPDFCommandData = IGdcKdMessageEnvelope<GdcKdCommandType.ExportToPDF, null>;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link isExportCommandData}.
```
Comment (lanhq) [inline src/embedding/kd.ts:439]: `{@link isExportCommandData}` => `{@link ExportToPDFCommandData}`

File: src/embedding/kd.ts [hunk 3]
```diff
@@ -308,6 +340,110 @@ export namespace EmbeddedKpiDashboard {
...
+     * Adds new attribute filter to filter bar and starts the filter customization flow.
+     *
+     * Contract:
+     *
+     * -  if KD is currently editing a dashboard, adds new attribute filter, starts customization flow; FilterAdded
+     *    will be posted right after customization starts
+     *
+     * -  if KD is currently in view mode or does not show any dashboard, will post CommandFailed
+     *
+     * @remarks use {@link AddFilterCommand} factory function to instantiate
```
Comment (lanhq) [inline src/embedding/kd.ts:408]: `factory function` is function to create an instance of `AddFilterCommand`. But here we didn't have this kind of function. So please remove this comment.
Please check and remove all others remarks

File: src/embedding/kd.ts [hunk 4]
```diff
@@ -308,6 +340,110 @@ export namespace EmbeddedKpiDashboard {
...
+     * -  if KD is currently editing a dashboard, then depending on widget type:
+     *    -  KPI is added to dashboard, customization flow is started, WidgetAdded will be posted
+     *    -  Insight is added to dashboard, WidgetAdded will be posted
+     *
+     * -  if insight reference included in command payload does not refer to a valid insight, CommandFailed
+     *    will be posted
+     *
+     * -  if KD is in view mode or not showing any dashboard, then CommandFailed will be posted
+     *
+     * @remarks use {@link AddWidgetCommand} to create instances of AddWidget
```
Comment (lanhq) [inline src/embedding/kd.ts:383]: remove this remarks comment

File: src/embedding/kd.ts [hunk 5]
```diff
@@ -308,6 +340,110 @@ export namespace EmbeddedKpiDashboard {
...
+    /**
+     * Exports dashboard to PDF.
+     *
+     * Contract:
+     *
+     * -  if KD shows dashboard in view mode, will export dashboard to PDF and post ExportFinished once ready for
+     *    exporting
+     * -  if KD shwows dashboard in edit mode or not not showing any dashboard, CommandFailed will
+     *    be posted
+     * @remarks use {@link ExportToPDFCommand} factory function to instantiate
```
Comment (lanhq) [inline src/embedding/kd.ts:432]: remove this remarks comment

File: src/embedding/kd.ts [hunk 6]
```diff
@@ -308,6 +340,110 @@ export namespace EmbeddedKpiDashboard {
...
+     * -  if KD is currently in view mode or does not show any dashboard, will post CommandFailed
+     *
+     * @remarks use {@link AddFilterCommand} factory function to instantiate
+     */
+    export type AddFilterCommand = IGdcKdMessageEvent<GdcKdCommandType.AddFilter, null>;
+
+    export type AddFilterCommandData = IGdcKdMessageEnvelope<GdcKdCommandType.AddFilter, null>;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link isAddFilterCommandData}.
```
Comment (lanhq) [inline src/embedding/kd.ts:415]: `{@link isAddFilterCommandData}` => `{@link AddFilterCommandData}`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/99/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/184/ : SUCCESS in 56s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/100/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/185/ : SUCCESS in 24s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/101/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/186/ : SUCCESS in 53s

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-gate/34/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/187/ : SUCCESS in 23s
- https://checklist.intgdc.com/job/compliance/job/gooddata-typings-fossa-scanning-zuul-el7/2/ : SUCCESS in 26s (non-voting)

**Pull request merged**